### ISSUE TITLE
Implement Accessibility (a11y) recommendations

### ISSUE THREAD

**Issue by bobbingwide** [bobbingwide/oik#79]
### Requirement
Implement recommendations as documented in:
[Web Content Accessibility Guidelines (WCAG) 2.0](https://www.w3.org/TR/WCAG20/)

Start with Conformance level A.

See also:
- [Web Accessibility Initiative](https://www.w3.org/WAI/)
- [W3C Accessibility Standards]( https://www.w3.org/standards/webdesign/accessibility)

### Proposed solution
- Implement recommendations where the oik plugin is believed to be non-conformant.
- Implement solutions to enable web pages that incorporate output from the oik base plugin to comply with Level A conformance.

### Notes

- The low level functions delivered in the oik base plugin are expected to generate valid HTML.
- Some of the functions generate tags with attributes which aren't compliant with the latest a11y guidelines.

**Comment by bobbingwide** [comment]
Reply from Graham Armfield regarding title attribute on links.
Hi Herb. If the title attribute just repeats the link text, then yes they are superfluous and lead to over announcing in some screen readers.

If the title attribute informs user that link opens new window or tab then according to WCAG techniques that's a valid use. However, although the title MAY be available to screen reader users when they tab to a link, the value of the title attribute is not exposed in the 'list links' facility that desktop screen readers have, and which people do use a lot. So that's not great. Also, some screen readers do not announce title attributes on links, either by default or because of user settings.

Also, of course, title attributes are (by default) only seen by sighted users who are using a mouse. Not by sighted keyboard users, or users using touch input. So warning users that links open new window/tab by just using the title attribute is not the right answer either - as this affects more than just blind people.

So to sum up: In my view using the title attribute is never a good idea - except when it's used to label the contents of an <iframe>.

**Comment by bobbingwide** [comment]
### Re: `<abbr>` and `<acronym>`
Also from Graham on the `<abbr>` tag.

Sadly, very few screen readers support the <abbr> tag in a meaningful way (I think just JAWS). So best practice with abbreviations and acronyms is to define them once on first use on a page. Eg:

Three letter acronym (TLA) and then you can use the TLA or abbreviation subsequently.

Not an ideal situation I know.

### Proposed solution
- Change functions that deal with `<abbr>` and `<acronym>` to only produce mark up first time its used. 
- Reduce calls to these functions to a minimum.

**Comment by bobbingwide** [comment]
Some a11y changes have been implemented in the fixes for #81

**Comment by bobbingwide** [comment]
One of the easiest changes to make is to update retlink() so that it only adds the title= attribute if it's different from the text in the link.

**Comment by bobbingwide** [comment]
It's probably also necessary to remove the `title=` attribute from img tags
e.g. `<img class="bw_portfolio" src="https://qw/src/post-title/" title="post title" alt="post title"  />`

**Comment by bobbingwide** [comment]
> It's probably also necessary to remove the title= attribute from img tags

Well, it's been a long time since I looked at accessibility of oik's code. 
Having run WAVE against oik-plugins.com I thought it was time to add an `alt` attribute to the Featured image field displayed by the `[bw_field]` shortcode.  

I changed `bw_fields_get_featured_image()` to access the alt text for the featured image (using the same code as in `wp_get_attachment_image()` ), then passed this to `retimage()`.

```
$alt = trim( strip_tags( get_post_meta( $thumbnail_id, '_wp_attachment_image_alt', true ) ) );
$thumbnail = retimage( "featured", $thumbnail[0], $alt );
```


This worked but I got the WAVE Alert: _Redundant title text_ due to having both `title` and `alt` attributes.

So I need to
- change `retimage()` as well
- or call `wp_get_attachment_image` instead.

**Comment by bobbingwide** [comment]
> This worked but I got the WAVE Alert: Redundant title text due to having both title and alt attributes.

I got a similar problem with the `[bw_logo]` shortcode.
```
[bw_logo link='/' width=216 height=149]
```

**Comment by bobbingwide** [comment]
The `BW_::bw_select()` method's output causes WAVE to produce an Alert: Select missing label

Google Chrome's inspect also moans with "Incorrect use of `<label for="form element">`

<img width="411" height="84" alt="Image" src="https://github.com/user-attachments/assets/55231fea-11c6-4f55-b1d2-3a02759e30c1" />

The HTML causing the problem is this.
```
<label for="load_plus">
Add a plugin to be loaded
</label>
<select name="load_plus" fdprocessedid="mgxza8">
<option value"0"> None
</option>
...
</select>
```

**Comment by bobbingwide** [comment]
A similar problem occurs for textarea fields displayed by the `iarea()` function, which is called by `bw_textarea()` or `BW_::bw_textarea`

**Comment by bobbingwide** [comment]
In oik-options ( https://s.b/wordpress/wp-admin/admin.php?page=oik_menu ) there are several contrast errors and alerts.

- Contrast errors in Toggle panel screen reader text. one for each panel.
- Alert: Missing first level heading. 
- Alert: Layout table.

Need to check how it's done in other wp-admin pages.

**Comment by bobbingwide** [comment]
There's a Contrast error in the **postbox** that oik uses to simulate the Dashboard widgets.
It's been a bit funny since WordPress 4.7 or so. See comments in #81

In the WordPress dashboard the HTML for the postbox is different from the HTML generated by oik's
```
<div id="dashboard_right_now" class="postbox ">
<div class="postbox-header">
<h2 class="hndle ui-sortable-handle">At a Glance</h2>
<div class="handle-actions hide-if-no-js">

Two buttons that I didn't implement. 

<button type="button" class="handle-order-higher" aria-disabled="true" aria-describedby="dashboard_right_now-handle-order-higher-description" fdprocessedid="tifi7e"><span class="screen-reader-text">Move up</span><span class="order-higher-indicator" aria-hidden="true"></span></button><span class="hidden" id="dashboard_right_now-handle-order-higher-description">Move At a Glance box up</span>
<button type="button" class="handle-order-lower" aria-disabled="false" aria-describedby="dashboard_right_now-handle-order-lower-description" fdprocessedid="2ehmze"><span class="screen-reader-text">Move down</span><span class="order-lower-indicator" aria-hidden="true"></span></button><span class="hidden" id="dashboard_right_now-handle-order-lower-description">Move At a Glance box down</span>

<button type="button" class="handlediv" aria-expanded="true" fdprocessedid="ws2pk">
<span class="screen-reader-text">Toggle panel: At a Glance</span>
<span class="toggle-indicator" aria-hidden="true"></span></button>
</div>
</div>
<div class="inside">
<div class="main">

Content of the postbox goes here.
	
</div>
</div>
</div>
```

**Comment by bobbingwide** [comment]
In the oik postbox the HTML is

```
<div class="postbox " id="oik_shortcode_options">
<button type="button" class="handlediv" aria-expanded="true" fdprocessedid="uxkdb5">
<span class="screen-reader-text">Toggle panel: Shortcode options</span>
<span class="toggle-indicator" aria-hidden="true">
</span>
</button>
<h3 class="hndle ui-sortable-handle">Shortcode options</h3>
<div class="inside">

Contents of the postbox goes here.
</div>
</div>

**Comment by bobbingwide** [comment]
Well I never... 
- Yesterday I didn't get any contrast errors on the dashboard page for the **Toggle panel:** button
- Well, I'm pretty sure I didn't
- but now I've got 32 contrast errors.
- 2 are for the **At a Glance** postbox: Move down, Toggle panel but NOT Move up!
- 0 for **Quick Draft**: - since it's hidden
- 3 for **WordPress Event and News**: Move up, Move down and Toggle panel.
- etc

<img width="1920" height="1032" alt="Image" src="https://github.com/user-attachments/assets/5fda5668-826f-4fb4-b8d3-36d1abafb437" />

**Comment by bobbingwide** [comment]
I believe the reason the **Move up** button wasn't a color contrast issue is because it's `aria-disabled="true"`.

**Comment by bobbingwide** [comment]
For more detail of a11y changes associated with oik's implementation of postbox see #81

**Comment by bobbingwide** [comment]
I believe the Color Contrast problems associated with postbox can be fixed with the following CSS.

```
/**
Attempted a11y improvement to avoid Color Contrast messages from WAVE for oik's dashboard postboxes
 */
button.handlediv span.screen-reader-text {
    color: #000000;
}
```

For the time being I've added this code to `oik.css`, 
but it would be better if core implemented this to avoid messages on the dashboard.

**Comment by bobbingwide** [comment]
After asking in #accessibility about the Color Contrast messages in the dashboard I asked ChatGPT how to identify text identified for a screen reader. 

> how can the WAVE tool be informed that text is screen reader text which won't be displayed to a sighted user so doesn't need a color contrast check performed against it?

It replied

- - - 
WAVE cannot be “told” to skip contrast checks on text that is present in the DOM.
If it’s in the markup, WAVE treats it as visible text, even if your CSS hides it visually or positions it off-screen.

However, there are correct ways to mark text as “screen-reader only”, and if you use these patterns, WAVE will automatically treat the text as hidden and it will not run color contrast checks.

Use a proper visually hidden utility class — the industry-standard pattern is this:

CSS

```
.visually-hidden {
  position: absolute !important;
  width: 1px !important;
  height: 1px !important;
  margin: -1px !important;
  padding: 0 !important;
  overflow: hidden !important;
  clip: rect(0, 0, 0, 0) !important;
  white-space: nowrap !important;
  border: 0 !important;
```

HTML
```
<span class="visually-hidden">Text for screen readers only</span>
```

**Comment by bobbingwide** [comment]
> However, there are correct ways to mark text as “screen-reader only”, and if you use these patterns, WAVE will automatically treat the text as hidden and it will not run color contrast checks.

I tried this and it would appear that ChatGPT has lied ( aka hallucinated ). 
- Yes, it might be true that the text isn't visible. 
- But the technique doesn't appear to work.
- WAVE's documentation in https://wave.webaim.org/help says

### How does WAVE handle "hidden" elements?
WAVE detects accessibility issues in all page elements, even those that are hidden using CSS, the hidden attribute, aria-hidden="true", and/or tabindex="-1". This is by design. In most cases, hidden elements are revealed to users at some point. We believe it's important that these potential end user accessibility issues in drop-down menus, tab panels, dialog boxes, carousels, and other "hidden" elements be identified.

In most cases the WAVE interface will provide an indication of icons that are relevant to hidden elements—the icon in the sidebar will be slightly dimmed and will present an informative tooltip. These elements may be examined by turning off the Styles option in the WAVE sidebar.

If WAVE has detected an error in an element that is hidden and that is never presented to the end user, or that is perhaps modified to address the error before being presented (something WAVE cannot detect), the WAVE error can be ignored - or you may instead choose to adapt the code to ensure the element is accessible in its hidden state.

In short, if we err on this point, we err on the side of users with disabilities. In other words, we have chosen to identify issues in hidden content that may be presented to users, rather than suppress these errors in cases where advanced development techniques have been implemented to potentially hide that content from users.

**Comment by bobbingwide** [comment]
> or you may instead choose to adapt the code to ensure the element is accessible in its hidden state.

For oik I'll ensure the element is accessible by setting the color to black `#000000`.

**Comment by bobbingwide** [comment]
I wrote to WAVE, usingtheir contact form
- - - 

Hi, I've been trying to improve the accessibility of my WordPress plugins. I have several Color Contrast errors which are associated with hidden fields which are classed as screen-reader-text. 

I asked ChatGPT "how can the WAVE tool be informed that text is screen reader text which won't be displayed to a sighted user so doesn't need a color contrast check performed against it?"  

It replied "WAVE can't be told to skip contrast checks, and it treats is as visible text even if your CSS hides it visually or positions it off the screen" 

It then went on to tell me that there are correct ways to mark text as "screen-reader only". And gave an example using class="visually-hidden".

I tried this and wasn't surprised to find that ChatGPT was lying / hallucinating. 

I've found your documentation on handling "hidden" elements which appears to confirm the result.

So my questions are:

1. Can you confirm ChatGPT is talking a load of rubbish with the example it gave me. See the URL.

2. Should I attempt to deal with Color Contrast messages related to screen-reader-text by, for example, setting foreground and background colors? OR will that hide a problem waiting in the wings. 

My feeling is that I should fix the problem even though the field is never expected to be seen.

thanks in advance for any information. Herb

**Comment by bobbingwide** [comment]
> etc

I decided to extend my CSS fixes for Color Contrast messages in the WordPress core dashboard.

<img width="532" height="58" alt="Image" src="https://github.com/user-attachments/assets/0207a16f-6227-4b53-9560-860beb6cc914" />

- This section had problems with screen reader text
- And the vertical bar separators were `#dcdcde` on white
- `.community-events-footer` needs to be `#74747b` for WCAG 2.1 AA
- or remove the split vertical bar?

**Comment by bobbingwide** [comment]
Each comment had 5 Color Contrast errors.

<img width="380" height="278" alt="Image" src="https://github.com/user-attachments/assets/46b47072-5af2-4b73-abe7-229db62441e5" />

- The 5 vertical bar separators in `.row-actions` were `#a7aaad`  on a light grey background `#f6f7f7`
- They need to be `#6e7177` for WCAG AA ( 4.55:1 )

**Comment by bobbingwide** [comment]
<img width="380" height="186" alt="Image" src="https://github.com/user-attachments/assets/bc49a697-6526-4a2f-846c-b2aa85299e50" />

- The 5 vertical bar separators in `.activity-block .subsubsub li` were `#dcdcde` on `#ffffff`
- They need to be `#74747b` for WCAG AA

**Comment by bobbingwide** [comment]
Latest WAVE. All the Color Contrast errors fixed with just a few lines of CSS, which I added to `oik.css`

<img width="1913" height="968" alt="Image" src="https://github.com/user-attachments/assets/5a562edb-9e21-4dbc-9911-f2b6aab02ba7" />

```
/**
  Attempted a11y improvement to avoid Color Contrast messages from WAVE for oik's dashboard postboxes.
  These also target the core dashboard items.
 */
button.handlediv span.screen-reader-text,
button.handle-order-lower span.screen-reader-text,
button.handle-order-higher span.screen-reader-text,
.community-events-footer .screen-reader-text {
    color: #000000;
    background-color: #ffffff;
}

.community-events-footer {
    color: #74747b;
}
.row-actions {
    color: #6e7177;
}

.activity-block .subsubsub li {
    color: #74747b;
}
```

**Comment by bobbingwide** [comment]
I got a reply from WebAIM regarding my WAVE / ChatGPT questions.

- - - 

 Herb -

Thank you for the inquiry. It’s correct that WAVE can’t be told to not check certain elements. It seems ChatGPT then gave guidance for hiding content visually but still allowing it to be accessible to screen reader users. While this can be a useful technique (though should be used with discretion), this still doesn’t hide the content from WAVE testing.

Our approach is to test hidden content because most often hidden content might later be presented to users. The best solution to these contrast errors is to define foreground and background colors that meet the WCAG requirements. At a technical level WCAG requires this even for elements that are not made visible.

Thanks,

Jared Smith
WebAIM


> The best solution to these contrast errors is to define foreground and background colors that meet the WCAG requirements.

- This is what I've done for the `.screen-reader-text` CSS. 
- It's may also be necessary to do this for the other CSS
- It depends on the background colours being used in other Administration Color Schemes.
- I'd only made changes for default.
- There are seven other (core) Color Schemes to consider. 


<img width="1622" height="294" alt="Image" src="https://github.com/user-attachments/assets/5a01d9bd-f6ae-4abc-afbe-a28561158a48" />

**Comment by bobbingwide** [comment]
> There are seven other (core) Color Schemes to consider.

With the Light color scheme I get 10 Color contrast errors.

<img width="1920" height="1020" alt="Image" src="https://github.com/user-attachments/assets/caed57b0-ef01-4350-8944-26183070c9a4" />

**Comment by bobbingwide** [comment]
I discussed the above in the #accessibility channel and was advised to raise an issue.
I raised it as https://core.trac.wordpress.org/ticket/64313
There's been some discussion.
It looks like my TRAC will be closed in favor of https://core.trac.wordpress.org/ticket/64308

**Comment by bobbingwide** [comment]
Moving on... another problem reported by WAVE is associated with checkboxes without labels in tables.
- I first encountered this with oik-unloader's `[active_plugins]` shortcode.
- For that plugin I prototyped a solution using a local method `icheckbox()`.
- My solution was to append an `aria-label` attribute that passes the `$name` parameter as the label value.
- This satisfies WAVE but may be quite useless for screen readers.

```
/**
 * Attempt to please the WAVE a11y tool.
 */
$it .= kv("aria-label", $name);
$it.= "/>";
```

Anyway, I'll try the same fix with the `icheckbox()` function in the bobbforms.php shared library.

**Comment by bobbingwide** [comment]
In several places I've used `bw_tablerow()` to display form fields in a table. 
- WAVE detects these tables as  Layout tables and produces an Alert for each.
- For tables which have headings we can update the call to `bw_tablerow()` that produces the header by adding two additional parameters: 'tr' and 'th'.
- For tables which don't have headers it makes sense to convert these to use grid layouts. 
- The underlying functions that use `bw_tablerow()`, eg `bw_checkbox()` already have the capability to to call `bw_gridrow()`
- In order to style the grids we'll need some CSS 
- and, if required, pass appropriate classnames in `$args['#class#]`.

**Comment by bobbingwide** [comment]
The `[bw_pages]` shortcode provides a `format` parameter which enables the user to specify which fields should be displayed.
A common use case, that I've use on oik-plugins is `format="IL"` or `"format="FL"`.
- I and F produce an image linked to the post
- L produces a link to the post
- WAVE reports this with a _Redundant link_ Alert
- If there are enough of these Alerts it can prevent the AIM score from reaching 10.

<img width="596" height="276" alt="Image" src="https://github.com/user-attachments/assets/2f660a7b-0032-48c8-8a38-f2d9e16633c8" />

### Proposed solution

- Add two new format values J and K
- J is for image and link and would replace "IL" or "FL"
- K is for link and image and would replace "LI" or "LF"

The generated HTML would only contain one link with the image and text arranged in child tags.
Note: This requires `wpautop()` processing to be disabled.  

### format="J"
<img width="468" height="203" alt="Image" src="https://github.com/user-attachments/assets/4cb3c805-03cf-4d22-89d3-fcba37963812" />

### format="K"

<img width="474" height="219" alt="Image" src="https://github.com/user-attachments/assets/08a37dd6-2f3a-4113-a64e-626c932f2fbb" />

If there is no featured or attached image the image will be omitted.

**Note:** For backward compatibility `[bw_pages]` shortcode expansion will not automatically convert the format parameter values.
Although this is quite possible, and can probably be achieved with a filter hook for `oik_shortcode_atts`.

**Comment by bobbingwide** [comment]
On revisiting s.b/oikcom/blog I encountered 4 errors due to the new shortcode examples I'd created to test the changes for `format="J"` and `format="K"` 

<img width="1176" height="900" alt="Image" src="https://github.com/user-attachments/assets/f5a5a940-be9f-4a1a-9d78-42d82f51b43e" />

- The generated HTML appears to have empty paragraphs.
- I'd eliminated these from the shortcode example pages by setting _Disable automatic paragraph generation_ to true.
- I dont know why they're still there. 
- As a workaround I should create excerpts for these shortcode examples.

**Comment by bobbingwide** [comment]
> I dont know why they're still there.

It seems that `wpautop()` was still being run for `the_excerpt`. 
I updated oik-css to fix this.

**Comment by bobbingwide** [comment]
On quite a few pages on `s.b/oikcom` I get an Alert: Suspicious link text for the output displaying the syntax for a shortcode that includes the parameter name `link`.

```
<a href="/oik_sc_param/nivo-link-parameter" title="nivo link parameter">link</a>
```

ChatGPT suggested a method of using the `aria-label` attribute in the `<a>` tag. 
```
<a href="/oik_sc_param/nivo-link-parameter" aria-label="nivo parameter">link</a>
```

It says this solution's Result is
- Sighted users see: link
- Screen readers announce: “nivo parameter, link”
- WAVE warning disappears

It added, with big ticks:

✔ WCAG 2.4.4 compliant
✔ No visual change
✔ Acceptable when visible text is constrained by design or UI rules

For WCAG 2.4.4 it probably meant Success Criterion ( SC ) 2.4.4 
see https://www.w3.org/WAI/WCAG22/Understanding/link-purpose-in-context.html

**Comment by bobbingwide** [comment]
> WAVE warning disappears

I made the change but the WAVE warning didn't disappear.

<img width="1353" height="695" alt="Image" src="https://github.com/user-attachments/assets/5cbd8e41-f8b8-4ee1-8689-f2bac321bd93" />

ChatGPT's second option was to use hidden screen reader text eg

```
<a href="/oik_sc_param/nivo-link-parameter">
  link
  <span class="screen-reader-text"> shortcode parameter</span>
</a>
```

- This solution worked.... but I still had the ARIA label.
- I removed the ARIA label and it still worked.
- Note: the CSS for class `screen-reader-text` is defined in `wp-block-library-inline-css`
